initial project

Arash-Fattani opened the pull request.

Arash-Fattani committed:
initial project

smzerehpoush commented on pom.xml:
```diff
@@ -0,0 +1,145 @@
+    <modelVersion>4.0.0</modelVersion>
+    <parent>
+        <groupId>org.springframework.boot</groupId>
+        <artifactId>spring-boot-starter-parent</artifactId>
```
always use latest LTS version.
from here: https://mvnrepository.com/artifact/org.springframework.boot/spring-boot-starter-parent

Arash-Fattani replied:
این رو نفهمیدم چکار کنم یعنی میگی modelVersion رو بذارم آخرین ورژنی که این سایت اومده؟ 
سایته تا 3.3.2 گذاشته ولی من اون رو گذاشتم ارور داد
![image](https://github.com/user-attachments/assets/8453009d-6874-44f2-95ff-1467c78464c8)

smzerehpoush replied:
ریلود کن. mvn clean compile بزن که دانلود شه اگر اتوماتیک نشد.

Arash-Fattani replied:
https://stackoverflow.com/questions/19759338/why-modelversion-of-pom-xml-is-necessary-and-always-set-to-4-0-0

Arash-Fattani replied:
https://maven.apache.org/guides/introduction/introduction-to-the-pom.html
 اینم از سایت خود maven

smzerehpoush commented on pom.xml:
```diff
@@ -0,0 +1,145 @@
+        <version>3.4.0-M1</version>
+        <relativePath/> <!-- lookup parent from repository -->
+    </parent>
+    <groupId>com.example.millieprojects.initialproject</groupId>
```
change group id and folder names to
gold.milli.initial-project

smzerehpoush commented on pom.xml:
```diff
@@ -0,0 +1,145 @@
+        <relativePath/> <!-- lookup parent from repository -->
+    </parent>
+    <groupId>com.example.millieprojects.initialproject</groupId>
+    <artifactId>financialtransactionsystem</artifactId>
```
rename to initial-project

smzerehpoush commented on pom.xml:
```diff
@@ -0,0 +1,145 @@
+    <name>financialtransactionsystem</name>
+    <description>Demo project for Spring Boot</description>
+    <url/>
+    <licenses>
```
remove unused blocks like licenses and developers ...

smzerehpoush commented on pom.xml:
```diff
@@ -0,0 +1,145 @@
+    <dependencies>
+        <dependency>
+            <groupId>org.springframework.boot</groupId>
+            <artifactId>spring-boot-starter-data-jdbc</artifactId>
```
وقتی starter-data داریم این نیاز نمیشه ها

smzerehpoush commented on pom.xml:
```diff
@@ -0,0 +1,145 @@
+        <dependency>
+            <groupId>org.projectlombok</groupId>
+            <artifactId>lombok</artifactId>
+            <version>1.18.34</version>
```
it's better to define versions in block

smzerehpoush commented on pom.xml:
```diff
@@ -0,0 +1,145 @@
+        <url/>
+    </scm>
+    <properties>
+        <java.version>21</java.version>
```
برای نزدیک تر شدن به شرایط واقعی از جاوا ۱۷ استفاده کن.

smzerehpoush commented on pom.xml:
```diff
@@ -0,0 +1,145 @@
+        </plugins>
+    </build>
+    <repositories>
+        <repository>
```
اینا اضافین. حذفشون کن

smzerehpoush commented on src/main/java/com/example/millieprojects/initialproject/financialtransactionsystem/controller/AccountController.java:
```diff
@@ -0,0 +1,56 @@
+
+@RestController
+public class AccountController {
+    private AccountService accountService;
```
use final variables and @RequiredArgsConstructor

Arash-Fattani committed:
almost finished pom.xml reviews

Arash-Fattani committed:
forgot one of the reviews

Arash-Fattani committed:
added RequiredArgsConstructor and final variables

smzerehpoush commented on src/main/java/com/example/millieprojects/initialproject/financialtransactionsystem/controller/AccountController.java:
```diff
@@ -0,0 +1,56 @@
+        this.accountService = accountService;
+    }
+
+    @Transactional
```
اینکه در اینجا چرا بهش نیاز هست یا نه رو باید فکر کنی بهش.
ولی اگر هم نیاز باشه ما در لایه پایین تر قرار میدیمش.

smzerehpoush commented on src/main/java/com/example/millieprojects/initialproject/financialtransactionsystem/controller/AccountController.java:
```diff
@@ -0,0 +1,56 @@
+            summary = "create a new Account",
+            description = "create a new account for a logged in account"
+    )
+    @PostMapping("/user/{id}/create-account")
```
اگر مدل RESTFul api رو مطالعه کنی میبینی که در اونجا verb نداریم. و با نوع متد حرفمون رو میزنیم مگر اینکه نشه.

smzerehpoush commented on src/main/java/com/example/millieprojects/initialproject/financialtransactionsystem/controller/AccountController.java:
```diff
@@ -0,0 +1,56 @@
+    )
+    @PostMapping("/user/{id}/create-account")
+    public Account saveAccount(@RequestBody Account account, @PathVariable int id){
+      return accountService.saveAccount(account, id);
```
در ورودی باید از مدل های DTO استفاده کنی

Arash-Fattani replied:
واسه Account هنوز mapper تعریف نشده چون یه سری مشکل دارم سر دیتابیس فیچر بعدیه که تو یه برنچ دیگه تعریف شده

smzerehpoush commented on src/main/java/com/example/millieprojects/initialproject/financialtransactionsystem/controller/AccountController.java:
```diff
@@ -0,0 +1,56 @@
+            description = "create a new account for a logged in account"
+    )
+    @PostMapping("/user/{id}/create-account")
+    public Account saveAccount(@RequestBody Account account, @PathVariable int id){
```
همخوانی نام وجود نداره
createAccount vs saveAccount

smzerehpoush commented on src/main/java/com/example/millieprojects/initialproject/financialtransactionsystem/controller/AccountController.java:
```diff
@@ -0,0 +1,56 @@
+            description = "Update Account type and balance"
+    )
+    @PutMapping("/user/{Uid}/account/{id}")
+    public Account updateAccount(@PathVariable int Uid, @PathVariable int id, @RequestBody Account account){
```
نام ها در java camelCase هستن. لطفا به waring های IDE توجه که من خیلی حساسم بهشون.

smzerehpoush commented on src/main/java/com/example/millieprojects/initialproject/financialtransactionsystem/controller/UserController.java:
```diff
@@ -0,0 +1,76 @@
+    private UserUpdateMapper userUpdateMapper;
+
+    @Autowired
+    public UserController(UserService userService,
```
final properties and requiredArgsConstructor

Arash-Fattani replied:
این ورژن قدیمیه

smzerehpoush commented on src/main/java/com/example/millieprojects/initialproject/financialtransactionsystem/controller/UserController.java:
```diff
@@ -0,0 +1,76 @@
+@Tag(name = "User Management", description = "Operations related to user management")
+public class UserController {
+    private UserService userService;
+    private UserCreateMapper userMapper;
```
برای هر سرویس یک مپر ایجاد میشه و مپینگ های متفاوت در اون قرار داده میشه. UserServiceMapper کفایت میکنه

smzerehpoush commented on src/main/java/com/example/millieprojects/initialproject/financialtransactionsystem/controller/UserController.java:
```diff
@@ -0,0 +1,76 @@
+            summary = "create a user",
+            description = "Save a newly registered user"
+    )
+    public UserDTOCreate saveUser(@RequestBody UserDTOCreate userRequest) {
```
در convention تیمی ما برای مدل های خروجی از اسامی استفاده میشه که به *ResponseDto  منتهی میشه و برای مدل های ورودی از اسامی استفاده میشه که به *RequestDto منتهی میشه و در پکیج های جداگانه قرار داده میشه.

smzerehpoush replied:
CreateUserRequestDto
CreateUserResponseDto

smzerehpoush commented on src/main/java/com/example/millieprojects/initialproject/financialtransactionsystem/controller/UserController.java:
```diff
@@ -0,0 +1,76 @@
+    )
+    public UserDTOCreate saveUser(@RequestBody UserDTOCreate userRequest) {
+
+        User user = this.userMapper.UserCreateFromRequest(userRequest);
```
چرا از this استفاده میکنی ؟

smzerehpoush commented on src/main/java/com/example/millieprojects/initialproject/financialtransactionsystem/controller/UserController.java:
```diff
@@ -0,0 +1,76 @@
+    public UserDTOCreate saveUser(@RequestBody UserDTOCreate userRequest) {
+
+        User user = this.userMapper.UserCreateFromRequest(userRequest);
+        return this.userMapper.UserDTOCreateFromUser(userService.saveUser(user));
```
بهتره که log کنی که چنین یوزری با این آیدی ایجاد شد.

Arash-Fattani replied:
log کردن فاز بعدی هست آخه

smzerehpoush commented on src/main/java/com/example/millieprojects/initialproject/financialtransactionsystem/controller/UserController.java:
```diff
@@ -0,0 +1,76 @@
+            description = "fetch all of the registered users"
+    )
+    public List<User> getUsers() {
+        return userService.getUsers();
```
نباید مدل خام برگردونی.
همیشه باید DTO برگردونی

smzerehpoush commented on src/main/java/com/example/millieprojects/initialproject/financialtransactionsystem/controller/UserController.java:
```diff
@@ -0,0 +1,76 @@
+        try {
+            User user = this.userMapper.UserCreateFromRequest(this.userUpdateMapper.UserUpdateFromRequest(userDTOUpdate));
+            return this.userUpdateMapper.UserDTOUpdateFromUser(
+                    this.userMapper.UserDTOCreateFromUser(userService.updateUser(user, id)));
```
این رو اصلا نمی فهمم

smzerehpoush commented on src/main/java/com/example/millieprojects/initialproject/financialtransactionsystem/entity/Account.java:
```diff
@@ -0,0 +1,37 @@
+import jakarta.persistence.*;
+import lombok.*;
+
+@Entity(name = "Accounts")
```
it's better to table name style is snake case

smzerehpoush commented on src/main/java/com/example/millieprojects/initialproject/financialtransactionsystem/entity/Account.java:
```diff
@@ -0,0 +1,37 @@
+    @Id
+    @GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "account_seq")
+    @SequenceGenerator(name = "account_seq", allocationSize = 1)
+    private int accountId;
```
id is a better name

smzerehpoush commented on src/main/java/com/example/millieprojects/initialproject/financialtransactionsystem/entity/Account.java:
```diff
@@ -0,0 +1,37 @@
+    @Column(unique = true)
+    private String accountNumber;
+    @NonNull
+    private double balance;
```
double is not good for financial calculations. use BIGDECIMAL instead

smzerehpoush commented on src/main/java/com/example/millieprojects/initialproject/financialtransactionsystem/entity/Account.java:
```diff
@@ -0,0 +1,37 @@
+@NoArgsConstructor
+@AllArgsConstructor
+@EqualsAndHashCode(of = "accountId")
+public class Account {
```
هدف از ایجاد این مدل چی بوده؟

Arash-Fattani replied:
منظورت از مدل Account هستش؟

smzerehpoush commented on src/main/java/com/example/millieprojects/initialproject/financialtransactionsystem/entity/Account.java:
```diff
@@ -0,0 +1,37 @@
+import lombok.*;
+
+@Entity(name = "Accounts")
+@Data
```
بهتره که این روی entity ها استفاده نشه

Arash-Fattani replied:
منظورت name هستش برای دیتابیس یا @Data ؟

smzerehpoush commented on src/main/java/com/example/millieprojects/initialproject/financialtransactionsystem/entity/User.java:
```diff
@@ -0,0 +1,50 @@
+
+import java.util.List;
+
+@Data
```
از این استفاده نشود

smzerehpoush commented on src/main/java/com/example/millieprojects/initialproject/financialtransactionsystem/entity/UserDTOCreate.java:
```diff
@@ -0,0 +1,23 @@
+
+    private String username;
+
+    private String password;
```
هیچ وقت پسورد رو به عنوان خروجی بر نگردون!!

smzerehpoush commented on src/main/java/com/example/millieprojects/initialproject/financialtransactionsystem/entity/UserDTOCreate.java:
```diff
@@ -0,0 +1,23 @@
+@AllArgsConstructor
+@Builder(toBuilder = true)
+@NoArgsConstructor
+public class UserDTOCreate {
```
CreateUserRequestDto

smzerehpoush commented on src/main/java/com/example/millieprojects/initialproject/financialtransactionsystem/entity/UserDTOCreate.java:
```diff
@@ -0,0 +1,23 @@
+import java.util.List;
+@Data
+@AllArgsConstructor
+@Builder(toBuilder = true)
```
builder is anti-pattern
pls do not use it

smzerehpoush commented on src/main/java/com/example/millieprojects/initialproject/financialtransactionsystem/entity/UserDTOUpdate.java:
```diff
@@ -0,0 +1,16 @@
+@NoArgsConstructor
+@Getter
+@Setter
+public class UserDTOUpdate {
```
UpdateUserRequestDto

smzerehpoush commented on src/main/java/com/example/millieprojects/initialproject/financialtransactionsystem/entity/UserDTOUpdate.java:
```diff
@@ -0,0 +1,16 @@
+
+import lombok.*;
+
+@Data
```
وقتی @Data هست به بقیه نیازی نیست. برو توش ببین که همه این ها رو داره.

smzerehpoush commented on src/main/java/com/example/millieprojects/initialproject/financialtransactionsystem/mapper/UserCreateMapper.java:
```diff
@@ -0,0 +1,21 @@
+        @Mapping(source = "password", target = "password")
+        @Mapping(source = "email", target = "email")
+        @Mapping(source = "roles", target = "roles")
+        User UserCreateFromRequest (UserDTOCreate request);
```
وقتی نام در مبدا و مقصد یکسان باشند به این موارد نیاز نیست

smzerehpoush commented on src/main/java/com/example/millieprojects/initialproject/financialtransactionsystem/mapper/UserUpdateMapper.java:
```diff
@@ -0,0 +1,20 @@
+    @Mapping(target = "roles",  expression = "java(java.util.Collections.emptyList())")
+    @Mapping(source = "email", target = "email")
+    @Mapping(source = "username", target = "username")
+    UserDTOCreate UserUpdateFromRequest (UserDTOUpdate request);
```
اسم متد ها باید camelCase باشند

smzerehpoush commented on src/main/java/com/example/millieprojects/initialproject/financialtransactionsystem/service/AccountService.java:
```diff
@@ -0,0 +1,73 @@
+
+    @Override
+    @Transactional
+    public Account saveAccount(Account account, int id) {
```
```suggestion
    public Account saveAccount(Account account, int userId) {
```

smzerehpoush commented on src/main/java/com/example/millieprojects/initialproject/financialtransactionsystem/service/AccountService.java:
```diff
@@ -0,0 +1,73 @@
+    @Transactional
+    public Account saveAccount(Account account, int id) {
+        User owner = userService.findUserById(id);
+        Account newAccount = account.toBuilder().owner(owner).build();
```
use constructor not builder

Arash-Fattani replied:
دارم صرفا یه فیلدیش رو تغییر میدم به نظرم اومد از استفاده از سترها تمیزتره

smzerehpoush commented on src/main/java/com/example/millieprojects/initialproject/financialtransactionsystem/service/AccountService.java:
```diff
@@ -0,0 +1,73 @@
+    }
+
+    @Override
+    @SneakyThrows
```
do not use this!

smzerehpoush commented on src/main/java/com/example/millieprojects/initialproject/financialtransactionsystem/service/AccountService.java:
```diff
@@ -0,0 +1,73 @@
+    public void deleteAccount(int uid, int id) {
+        User owner = userService.findUserById(uid);
+        Optional<Account> accountHolder = accountRepository.findById(id);
+        if (accountHolder.isPresent()){
```
اگر متدی بسازی که به عنوان ورودی ایدی اکانت و یوزر رو بگیره و چک کنه همخوانی این ها رو لازم نیست این کد تکرار بشه.

smzerehpoush commented on src/main/java/com/example/millieprojects/initialproject/financialtransactionsystem/service/UserService.java:
```diff
@@ -0,0 +1,62 @@
+import java.util.Optional;
+
+@Service
+public class UserService implements UserServiceInterface {
```
در جاوا کانونشن نام گذاری به این شکل هست:
UserService: interface
UserServiceImpl: implemented class

smzerehpoush commented on src/main/java/com/example/millieprojects/initialproject/financialtransactionsystem/service/UserService.java:
```diff
@@ -0,0 +1,62 @@
+public class UserService implements UserServiceInterface {
+    UserRepository userRepository;
+
+    public UserService(UserRepository userRepository) {
```
required args constructor

smzerehpoush commented on src/main/java/com/example/millieprojects/initialproject/financialtransactionsystem/service/UserService.java:
```diff
@@ -0,0 +1,62 @@
+    }
+
+    @Override
+    @SneakyThrows
```
do not use this

smzerehpoush commented on src/main/java/com/example/millieprojects/initialproject/financialtransactionsystem/service/UserService.java:
```diff
@@ -0,0 +1,62 @@
+    }
+
+    @Override
+    @Transactional
```
اینا رو همه جا گذاشتی ها
الکی نباید ازش استفاده کنی
برو تحقیق کن که چرا و کجا باید از این استفاده کرد.

smzerehpoush commented on src/main/resources/application.properties:
```diff
@@ -0,0 +1,9 @@
+spring.application.name=financialtransactionsystem
```
change to .yml file

smzerehpoush commented on src/main/resources/application.properties:
```diff
@@ -0,0 +1,9 @@
+server.port=8181
+spring.datasource.url=jdbc:postgresql://localhost:5432/postgres
+spring.datasource.username=postgres
+spring.datasource.password=Millie
```
milli

smzerehpoush commented on src/main/java/gold/milli/initialproject/InitialProjectApplication.java:
```diff
@@ -0,0 +1,22 @@
+				description = "My first real spring boot project"
+		)
+)
+public class InitialProjectApplication {
```
MilliApplication

Arash-Fattani committed:
eliminated use of 'this'

Arash-Fattani committed:
moved @Transaction to service layer and deleted use of it on read methods and also coordinated method names with URLs

Arash-Fattani committed:
renamed files to follow java and team naming conventions

Arash-Fattani committed:
eliminated @SneakyThrows and @builder

Arash-Fattani committed:
finished reviews

Arash-Fattani committed:
i just found out the order of dependencies is important too what a beautiful language spring is (the User APIs work properly)

smzerehpoush commented on src/main/java/gold/milli/initialproject/controller/AccountController.java:
```diff
@@ -0,0 +1,49 @@
+
+@RestController
+@RequiredArgsConstructor
+public class AccountController {
```
بهتره که ابتدای همه روت ها /api/v1 باشه
برای این منظور میتونی سر کلاس از انوتیشن RequestMapping استفاده کنی

smzerehpoush commented on src/main/java/gold/milli/initialproject/controller/AccountController.java:
```diff
@@ -0,0 +1,49 @@
+@RestController
+@RequiredArgsConstructor
+public class AccountController {
+    private final AccountServiceImpl accountServiceImpl;
```
باید AccountService که اینترفیس هست رو اینجکت کنی نه پیاده سازیش.
اسپرینگ اومده که اینا رو loose couple کنه دیگه. تو مستقیما اکانت رو به پیاده سازی یوزر وابسته کردی

smzerehpoush commented on src/main/java/gold/milli/initialproject/controller/AccountController.java:
```diff
@@ -0,0 +1,49 @@
+            description = "create a new account for a logged in account"
+    )
+    @PostMapping("/user/{userId}")
+    public Account createAccount(@RequestBody Account account, @PathVariable int userId){
```
به نظرم یوز آیدی توی ورودی همه ای پی آی ها باشه پس به بالاسر کلاس منتقل بشه.
بادی باید با انوتیشن @Validate ولیدیت بشه

Arash-Fattani replied:
برای Get و Post که تو یکیش داریم تمام user ها رو برمیگردونیم و تو یکیش داریم یه یوزر جدید میسازیم id چه معنی ای داره؟
عملا واسه post ما به یوزر id نسبت ندادیم که بیایم تو request بررسیش کنیم

smzerehpoush commented on src/main/java/gold/milli/initialproject/controller/AccountController.java:
```diff
@@ -0,0 +1,49 @@
+            description = "Update Account type and balance"
+    )
+    @PutMapping("/user/{userId}/{accountId}")
+    public Account updateAccount(@PathVariable int userId, @PathVariable int accountId,
```
خروجی ت که باز dto نیست

Arash-Fattani replied:
DTO مربوط به اکانت تو یه برنچ دیگه است

smzerehpoush commented on src/main/java/gold/milli/initialproject/controller/UserController.java:
```diff
@@ -0,0 +1,68 @@
+@RestController
+@RequiredArgsConstructor
+@Tag(name = "User Management", description = "Operations related to user management")
+public class UserController {
```
add request mapping

smzerehpoush commented on src/main/java/gold/milli/initialproject/controller/UserController.java:
```diff
@@ -0,0 +1,68 @@
+@RequiredArgsConstructor
+@Tag(name = "User Management", description = "Operations related to user management")
+public class UserController {
+    private final UserServiceImpl userServiceImpl;
```
use interfaces not implementations

smzerehpoush commented on src/main/java/gold/milli/initialproject/controller/UserController.java:
```diff
@@ -0,0 +1,68 @@
+            summary = "create a user",
+            description = "Save a newly registered user"
+    )
+    public UserResponseDto createUser(@RequestBody CreateUserRequestDto userRequest) {
```
ببین برای چیز های جنرال میتونی از لفظ response استفاده نکنی
مثلا UserDto, AccountDto

ولی مثلا برای اونایی که عملیات دارن باید از این لفظ استفاده کنی

smzerehpoush commented on src/main/java/gold/milli/initialproject/controller/UserController.java:
```diff
@@ -0,0 +1,68 @@
+    )
+    public List<UserResponseDto> fetchAllUsers() {
+        return userServiceImpl.fetchAllUsers().stream()
+                .map(userServiceMapper.userResponseMapper::responseFromUser).collect(Collectors.toList());
```
بهتره که pagination اضافه کنی بهش و Page برگردونی

smzerehpoush commented on src/main/java/gold/milli/initialproject/controller/UserController.java:
```diff
@@ -0,0 +1,68 @@
+    public UserResponseDto updateUser(@RequestBody UpdateUserRequestDto updateUserRequestDto, @PathVariable Integer userId) {
+        try {
+            User user = userServiceMapper.updateUserRequestMapper.userFromUpdateRequest(updateUserRequestDto);
+            System.out.println(user);
```
این برای کد کنسوله.
ما لاگ میزنیم.
از @Slf4j استفاده کن

smzerehpoush commented on src/main/java/gold/milli/initialproject/controller/UserController.java:
```diff
@@ -0,0 +1,68 @@
+            System.out.println(user);
+            return userServiceMapper.userResponseMapper.responseFromUser(userServiceImpl.updateUser(user, userId));
+        } catch (Exception e) {
+            System.out.println(e.getMessage());
```
باید خطا به کاربر برگردونی

smzerehpoush commented on src/main/java/gold/milli/initialproject/entity/Account.java:
```diff
@@ -0,0 +1,39 @@
+    @Id
+    @GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "account_seq")
+    @SequenceGenerator(name = "account_seq", allocationSize = 1)
+    private int id;
```
Integer

smzerehpoush commented on src/main/java/gold/milli/initialproject/entity/Account.java:
```diff
@@ -0,0 +1,39 @@
+public class Account {
+    @Id
+    @GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "account_seq")
+    @SequenceGenerator(name = "account_seq", allocationSize = 1)
```
account_id_seq

smzerehpoush replied:
add columnName to all properties

smzerehpoush commented on src/main/java/gold/milli/initialproject/entity/Account.java:
```diff
@@ -0,0 +1,39 @@
+
+    @NonNull
+    @Column(unique = true)
+    @Setter(AccessLevel.NONE)
```
remove this

smzerehpoush commented on src/main/java/gold/milli/initialproject/entity/Account.java:
```diff
@@ -0,0 +1,39 @@
+    @SequenceGenerator(name = "account_seq", allocationSize = 1)
+    private int id;
+
+    @NonNull
```
no need to this

Arash-Fattani replied:
منظورت @NonNull هستش؟
چرا بهش نیازی نیست؟

smzerehpoush commented on src/main/java/gold/milli/initialproject/entity/CreateUserRequestDto.java:
```diff
@@ -0,0 +1,19 @@
+import java.util.List;
+@Data
+@AllArgsConstructor
+@ToString
```
no need to this when u have @Data

smzerehpoush commented on src/main/java/gold/milli/initialproject/entity/CreateUserRequestDto.java:
```diff
@@ -0,0 +1,19 @@
+
+    private String username;
+
+    private String password;
```
**WTF**

smzerehpoush commented on src/main/java/gold/milli/initialproject/entity/User.java:
```diff
@@ -0,0 +1,51 @@
+    @GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "user_seq")
+    @SequenceGenerator(name= "user_seq", sequenceName = "user_sequence", allocationSize = 1)
+    @Id
+    private int userId;
```
id

smzerehpoush commented on src/main/java/gold/milli/initialproject/entity/User.java:
```diff
@@ -0,0 +1,51 @@
+@EqualsAndHashCode(of = "userId")
+@Getter
+@Setter
+@ToString
```
do not use this in entities and write your own

smzerehpoush commented on src/main/java/gold/milli/initialproject/entity/User.java:
```diff
@@ -0,0 +1,51 @@
+//TODO: when the application's data increases use a higher value for allocation size to increase performance
+public class User {
+    @GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "user_seq")
+    @SequenceGenerator(name= "user_seq", sequenceName = "user_sequence", allocationSize = 1)
```
user_id_seq

smzerehpoush commented on src/main/java/gold/milli/initialproject/entity/User.java:
```diff
@@ -0,0 +1,51 @@
+    private String username;
+
+    @NonNull
+    @Getter(AccessLevel.NONE)
```
remove

smzerehpoush commented on src/main/java/gold/milli/initialproject/entity/User.java:
```diff
@@ -0,0 +1,51 @@
+    @ElementCollection
+    @Enumerated(EnumType.STRING)
+    @Column(name = "role")
+    @CollectionTable(name = "user_roles", joinColumns = @JoinColumn(name = "user_id"))
```
string to enum conversion is a better solution for this problem

smzerehpoush commented on src/main/java/gold/milli/initialproject/entity/User.java:
```diff
@@ -0,0 +1,51 @@
+    private List<Role> roles;
+
+    @OneToMany(mappedBy = "owner")
+    @JsonIgnore
```
remove this

smzerehpoush commented on src/main/java/gold/milli/initialproject/entity/User.java:
```diff
@@ -0,0 +1,51 @@
+
+
+    public void addAccount(Account account){
+        this.accounts.add(account);
```
check if account is not null

smzerehpoush commented on src/main/java/gold/milli/initialproject/mapper/CreateUserRequestMapper.java:
```diff
@@ -0,0 +1,12 @@
+
+
+@Mapper(componentModel = "spring")
+public interface CreateUserRequestMapper {
```
UserServiceMapper

smzerehpoush commented on src/main/java/gold/milli/initialproject/mapper/UpdateUserRequestMapper.java:
```diff
@@ -0,0 +1,11 @@
+import org.mapstruct.Mapper;
+
+@Mapper(componentModel = "spring")
+public interface UpdateUserRequestMapper {
```
move this to UserServiceMapper

smzerehpoush replied:
and also never map dto to entity
load it from database.

but map entity to dto is ok

smzerehpoush commented on src/main/java/gold/milli/initialproject/mapper/CreateUserRequestMapper.java:
```diff
@@ -0,0 +1,12 @@
+
+@Mapper(componentModel = "spring")
+public interface CreateUserRequestMapper {
+        User userFromCreateRequest(CreateUserRequestDto request);
```
naming should be to+{destinationType}
e.g. toUser

smzerehpoush commented on src/main/java/gold/milli/initialproject/mapper/UserResponseMapper.java:
```diff
@@ -0,0 +1,11 @@
+
+@Mapper(componentModel = "spring")
+public interface UserResponseMapper {
+    UserResponseDto responseFromUser(User request);
```
toUserDto

smzerehpoush commented on src/main/java/gold/milli/initialproject/mapper/UserServiceMapper.java:
```diff
@@ -0,0 +1,13 @@
+
+@RequiredArgsConstructor
+@Component
+public class UserServiceMapper {
```
no need to this class.
just create a mapper for each service and add methods for each conversion

smzerehpoush commented on src/main/java/gold/milli/initialproject/service/AccountService.java:
```diff
@@ -0,0 +1,15 @@
+public interface AccountService {
+    Account createAccount(Account account, int userId);
+
+    List<Account> fetchAllAccounts(int userId);
```
getUserAccounts

smzerehpoush commented on src/main/java/gold/milli/initialproject/service/AccountServiceImpl.java:
```diff
@@ -0,0 +1,68 @@
+public class AccountServiceImpl implements AccountService {
+
+    private final AccountRepository accountRepository;
+    private final UserServiceImpl userServiceImpl;
```
UserService

smzerehpoush commented on src/main/java/gold/milli/initialproject/service/AccountServiceImpl.java:
```diff
@@ -0,0 +1,68 @@
+        return accountRepository.save(prevAccount);
+    }
+
+    public Account checkAccount(User owner, Integer accountId) throws Exception {
```
getAccountAndCheckOwnership is better name

smzerehpoush commented on src/main/java/gold/milli/initialproject/service/UserService.java:
```diff
@@ -0,0 +1,17 @@
+
+    List<User> fetchAllUsers();
+
+    User updateUser(User user, Integer userId) throws Exception;
```
باید dto به عنوان ورودی بگیره نه خود ابجکت رو

Arash-Fattani replied:
تو لایه کنترلر اومدم مپ کردم به یوزر یعنی نیاز نیست کلا به مپ کردن؟

smzerehpoush commented on src/main/java/gold/milli/initialproject/service/UserServiceImpl.java:
```diff
@@ -0,0 +1,49 @@
+    @Override
+    public User findUserById(int userId){
+        Optional<User> userHolder =  userRepository.findById(userId);
+        return userHolder.orElse(null);
```
throw user not found exeption

smzerehpoush commented on src/main/resources/application.yml:
```diff
@@ -0,0 +1,22 @@
+        dialect: org.hibernate.dialect.PostgreSQLDialect
+  application:
+    name: milliApplication
+  profiles:
```
remove this

smzerehpoush reviewed (changes requested):
check again

Arash-Fattani committed:
add /api/v1 to all URLs
coupled interface instead of implementation
added validation to the project (for User DTOs)

Arash-Fattani committed:
- deleted print commands
- deleted primitive types
+ added pagination
- deleted @Getter, @Setter, @String from the entities
- deleted @NonNull
+ added column names
+ added hashing to password

Arash-Fattani committed:
forgot to push (the hash part)

Arash-Fattani committed:
finished reviews and tested all of the APIs for User

Arash-Fattani committed:
- deleted DTO to entity mapping
+ added URL accounts to the class RequestMapping

Arash-Fattani committed:
pom.xml small update

Arash-Fattani committed:
small changes in services (clean code)

Arash-Fattani committed:
- eliminated int throughout the project

Arash-Fattani committed:
+ added column name to accounts in Use

Arash-Fattani committed:
+ added AccountDto and CreateAccountRequestDto
checked and tested Get and Post APIs for account controller

Arash-Fattani committed:
+ updated req

Arash-Fattani committed:
- converted Integer id to Long

Arash-Fattani committed:
nothing important
